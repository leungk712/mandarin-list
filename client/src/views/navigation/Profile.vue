<template>
  <application-layout>
    <v-container class="container-border profile-container" v-if="userInfo">
      <h1>
        <v-icon large class="pb-1">
          person_outline
        </v-icon>
        My Profile
      </h1>
      <p>Member since {{ userSince }}</p>
      <v-divider />
      <v-row class="my-6">
        <v-col cols="12 py-0">
          <v-row class="ml-1 mb-2">
            <v-icon class="mr-1">analytics</v-icon>
            <h2>Statistics</h2>
          </v-row>
        </v-col>
        <v-col cols="6">
          <v-card class="mx-auto" color="#eee" max-width="600">
            <v-card-title>
              <v-icon large left>
                mdi-ideogram-cjk
              </v-icon>
              <span class="title font-weight-light">Characters</span>
            </v-card-title>

            <v-card-text
              data-testid="profile-characters-card-text"
              class="profile-characters-card-text display-1 font-weight-bold text-center"
            >
              {{ posts.mandarinList.length }} characters
            </v-card-text>
          </v-card>
        </v-col>
        <v-col cols="6">
          <v-card class="mx-auto" color="#26c6da" dark max-width="600">
            <v-card-title>
              <v-icon large left>
                mdi-format-list-bulleted
              </v-icon>
              <span class="title font-weight-light">Categories</span>
            </v-card-title>

            <v-card-text
              data-testid="profile-categories-card-text"
              class="profile-categories-card-text display-1 font-weight-bold text-center"
            >
              {{ categories.categoriesList.length }} categories
            </v-card-text>
          </v-card>
        </v-col>
        <v-col cols="6">
          <v-card class="mx-auto" color="#BA68C8" dark max-width="600">
            <v-card-title>
              <v-icon large left>
                mdi-book-open-variant
              </v-icon>
              <span class="title font-weight-light">Stories</span>
            </v-card-title>

            <v-card-text
              data-testid="profile-stories-card-text"
              class="profile-stories-card-text display-1 font-weight-bold text-center"
            >
              {{ stories.storiesList.length }} stories
            </v-card-text>
          </v-card>
        </v-col>
        <v-col cols="6">
          <v-card class="mx-auto" color="#F06292" dark max-width="600">
            <v-card-title>
              <v-icon large left>
                favorite_border
              </v-icon>
              <span class="title font-weight-light">Favorites</span>
            </v-card-title>

            <v-card-text
              data-testid="profile-favorites-card-text"
              class="profile-favorites-card-text display-1 font-weight-bold text-center"
            >
              {{ favorites }} favorites
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
    </v-container>
  </application-layout>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import { namespace, State } from "vuex-class";
import {
  CategoriesState,
  PostsState,
  StoriesState,
  UserData,
  UserState
} from "@/models";
import ApplicationLayout from "@/components/layouts/ApplicationLayout.vue";
import CategoriesModule from "@/store/modules/categories";
import PostsModule from "@/store/modules/posts";
import StoriesModule from "@/store/modules/stories";
import UserModule from "@/store/modules/user";
import moment from "moment";

const categories = namespace(CategoriesModule.name);
const posts = namespace(PostsModule.name);
const stories = namespace(StoriesModule.name);
const user = namespace(UserModule.name);

@Component({
  name: "Profile",
  components: { ApplicationLayout }
})
export default class Profile extends Vue {
  // ===== Store ===== //
  @State("categories") public categories!: CategoriesState;
  @State("posts") public posts!: PostsState;
  @State("stories") public stories!: StoriesState;
  @State("user") public user!: UserState;

  // ===== Data ===== //

  // ===== Methods ===== //

  // ===== Computed ===== //
  get userInfo(): UserData {
    return this.user.user!;
  }

  get userSince(): string {
    return moment(this.userInfo.date)
      .utc()
      .format("MMMM Do, YYYY");
  }

  get favorites() {
    return this.posts.mandarinList.filter(characters => characters.starred)
      .length;
  }

  // ===== Lifecycle Hooks ===== //
}
</script>

<style scoped>
.profile-container {
  width: 85%;
}
</style>
