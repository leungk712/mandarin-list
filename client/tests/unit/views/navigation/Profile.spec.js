import { createLocalVue, shallowMount } from "@vue/test-utils";
import { mockUser } from "../../mocks/mock-user";
import Profile from "@/views/navigation/Profile.vue";
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuetify);

const localVue = createLocalVue();
localVue.use(Vuex);

const elem = document.createElement("div");
if (document.body) {
    document.body.appendChild(elem);
};

document.body.setAttribute("data-app", true);

const actions = {};

const mockStore = {
    user: {
        state: {
            accessToken: "Bearer et2085ofgu",
            avatarColor: "#eeeeee",
            isLoggedIn: true,
            loadingState: [],
            user: mockUser
        },
        actions
    },
    posts: {
        state: {},
        actions
    }
};

describe('Profile', () => {
    let wrapper;
    const store = new Vuex.Store({
        modules: mockStore
    });
    const mountFunction = (options, store = mockStore) => {
        return shallowMount(Profile, {
           attachTo: elem,
           localVue,
           store,
           vuetify: new Vuetify(),
           ...options
        });
    };
    describe("template", () => {
        it("should match snapshot", () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.$el).toMatchSnapshot();
            wrapper.destroy();
        });
    });
    describe("script", () => {
        it('userInfo returns the correct info', () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.userInfo).toEqual(mockUser);
            wrapper.destroy();
        })
    })
});