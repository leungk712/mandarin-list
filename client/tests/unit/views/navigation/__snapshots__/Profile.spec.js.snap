// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile template should match snapshot 1`] = `
<application-layout-stub>
  <v-container-stub
    class="container-border profile-container"
    tag="div"
  >
    <h1>
      <v-icon-stub
        class="pb-1"
        large=""
      >
        
        person_outline
      
      </v-icon-stub>
      
      My Profile
    
    </h1>
     
    <p>
      Member since February 20th, 2021
    </p>
     
    <v-divider-stub />
     
    <v-row-stub
      class="my-6"
      tag="div"
    >
      <v-col-stub
        cols="12 py-0"
        tag="div"
      >
        <v-row-stub
          class="ml-1 mb-2"
          tag="div"
        >
          <v-icon-stub
            class="mr-1"
          >
            analytics
          </v-icon-stub>
           
          <h2>
            Statistics
          </h2>
        </v-row-stub>
      </v-col-stub>
       
      <v-col-stub
        cols="6"
        tag="div"
      >
        <v-card-stub
          class="mx-auto"
          color="#eee"
          loaderheight="4"
          maxwidth="600"
          tag="div"
        >
          <v-card-title-stub>
            <v-icon-stub
              large=""
              left=""
            >
              
              mdi-ideogram-cjk
            
            </v-icon-stub>
             
            <span
              class="title font-weight-light"
            >
              Characters
            </span>
          </v-card-title-stub>
           
          <v-card-text-stub
            class="profile-characters-card-text display-1 font-weight-bold text-center"
            data-testid="profile-characters-card-text"
            datatestid="profile-characters-card-text"
          >
            
            6 characters
          
          </v-card-text-stub>
        </v-card-stub>
      </v-col-stub>
       
      <v-col-stub
        cols="6"
        tag="div"
      >
        <v-card-stub
          class="mx-auto"
          color="#26c6da"
          dark="true"
          loaderheight="4"
          maxwidth="600"
          tag="div"
        >
          <v-card-title-stub>
            <v-icon-stub
              large=""
              left=""
            >
              
              mdi-format-list-bulleted
            
            </v-icon-stub>
             
            <span
              class="title font-weight-light"
            >
              Categories
            </span>
          </v-card-title-stub>
           
          <v-card-text-stub
            class="profile-categories-card-text display-1 font-weight-bold text-center"
            data-testid="profile-categories-card-text"
            datatestid="profile-categories-card-text"
          >
            
            3 categories
          
          </v-card-text-stub>
        </v-card-stub>
      </v-col-stub>
       
      <v-col-stub
        cols="6"
        tag="div"
      >
        <v-card-stub
          class="mx-auto"
          color="#BA68C8"
          dark="true"
          loaderheight="4"
          maxwidth="600"
          tag="div"
        >
          <v-card-title-stub>
            <v-icon-stub
              large=""
              left=""
            >
              
              mdi-book-open-variant
            
            </v-icon-stub>
             
            <span
              class="title font-weight-light"
            >
              Stories
            </span>
          </v-card-title-stub>
           
          <v-card-text-stub
            class="profile-stories-card-text display-1 font-weight-bold text-center"
            data-testid="profile-stories-card-text"
            datatestid="profile-stories-card-text"
          >
            
            3 stories
          
          </v-card-text-stub>
        </v-card-stub>
      </v-col-stub>
       
      <v-col-stub
        cols="6"
        tag="div"
      >
        <v-card-stub
          class="mx-auto"
          color="#F06292"
          dark="true"
          loaderheight="4"
          maxwidth="600"
          tag="div"
        >
          <v-card-title-stub>
            <v-icon-stub
              large=""
              left=""
            >
              
              favorite_border
            
            </v-icon-stub>
             
            <span
              class="title font-weight-light"
            >
              Favorites
            </span>
          </v-card-title-stub>
           
          <v-card-text-stub
            class="profile-favorites-card-text display-1 font-weight-bold text-center"
            data-testid="profile-favorites-card-text"
            datatestid="profile-favorites-card-text"
          >
            
            4 favorites
          
          </v-card-text-stub>
        </v-card-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</application-layout-stub>
`;
